File: packages/zynth-hypervisor/android/src/main/java/com/zynth/hypervisor/ZynthHypervisorModule.kt
package com.zynth.hypervisor

import com.zynth.kit.runtime.ZynthModule
import com.zynth.kit.runtime.ZynthRuntime
import com.zynth.kit.runtime.ZynthArgs
import org.json.JSONObject

class ZynthHypervisorModule(
    private val runtime: ZynthRuntime,
    private val onMessage: (Any?) -> Unit
) : ZynthModule {
    override val name = "ZynthHypervisor"

    override val exportedMethods: List<String> = listOf("postMessage")
    override val protectedMethods: List<String> = listOf("postMessage")

    override fun call(method: String, args: ZynthArgs): JSONObject {
        return when (method) {
            "postMessage" -> {
                // Try extracting from "message" key if it's a dict, or first index if it's an array
                val params = try { args.nestedAt(0) } catch (e: Exception) { null }
                val message = (try { params?.getString("message") } catch (e: Exception) { null }) 
                    ?: (try { args.getStringAt(0) } catch (e: Exception) { null })
                onMessage(message)
                JSONObject()
            }
            else -> throw IllegalArgumentException("Unsupported method: $method")
        }
    }
}

File: packages/zynth-hypervisor/ios/ZynthHypervisorModule.swift
import Foundation
import ZynthKit

@objc(ZynthHypervisorModule)
class ZynthHypervisorModule: NSObject, ZynthModule, ZynthSyncModule {
    var name: String = "ZynthHypervisor"
    
    var exportedMethods: [String] {
        return ["postMessage"]
    }

    var protectedMethods: [String] {
        return ["postMessage"]
    }
    
    private weak var runtime: ZynthRuntime?
    
    init(runtime: ZynthRuntime) {
        self.runtime = runtime
    }
    
    func call(method: String, args: ZynthArgs) throws -> Any? {
        switch method {
        case "postMessage":
            // Guest sending message to Host
            // Try extracting from "message" key if it's a dict, or first index if it's an array
            let message = (try? args.string("message")) ?? (try? args.getString(0))
            if let message = message {
                if let runtime = runtime {
                    NotificationCenter.default.post(
                        name: .didReceiveGuestMessage,
                        object: runtime,
                        userInfo: ["message": message]
                    )
                }
            }
            return nil
        default:
            throw ZynthModuleError.methodNotExported(module: name, method: method)
        }
    }
    
    func callSync(method: String, args: ZynthArgs) throws -> Any? {
        return nil
    }
}

extension Notification.Name {
    static let didReceiveGuestMessage = Notification.Name("ZynthHypervisorDidReceiveGuestMessage")
}
